libcudacxx/test/libcudacxx/std/numerics/c.math/logarithms.pass.cpp

```diff
@@ -0,0 +1,205 @@
+#if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+  static_assert((cuda::std::is_same<decltype(cuda::std::log2((long double) 0)), long double>::value), "");
+#endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+  static_assert((cuda::std::is_same<decltype(cuda::std::log2f(0)), float>::value), "");
+#if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+  static_assert((cuda::std::is_same<decltype(cuda::std::log2l(0)), long double>::value), "");
+#endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+#ifdef _LIBCUDACXX_HAS_NVFP16
+  static_assert((cuda::std::is_same<decltype(cuda::std::log2((__half) 0)), __half>::value), "");
+#endif // _LIBCUDACXX_HAS_NVFP16
+#ifdef _LIBCUDACXX_HAS_NVBF16
+  static_assert((cuda::std::is_same<decltype(cuda::std::log2((__nv_bfloat16) 0)), __nv_bfloat16>::value), "");
+#endif // _LIBCUDACXX_HAS_NVBF16
+  assert(cuda::std::log2(1) == 0);
+}
+
+__host__ __device__ void test_logb()
+{
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((float) 0)), float>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((bool) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((unsigned short) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((int) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((unsigned int) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((long) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((unsigned long) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((long long) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((unsigned long long) 0)), double>::value), "");
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((double) 0)), double>::value), "");
+#if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((long double) 0)), long double>::value), "");
+#endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+  static_assert((cuda::std::is_same<decltype(cuda::std::logbf(0)), float>::value), "");
+#if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+  static_assert((cuda::std::is_same<decltype(cuda::std::logbl(0)), long double>::value), "");
+#endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+#ifdef _LIBCUDACXX_HAS_NVFP16
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((__half) 0)), __half>::value), "");
+#endif // _LIBCUDACXX_HAS_NVFP16
+#ifdef _LIBCUDACXX_HAS_NVBF16
+  static_assert((cuda::std::is_same<decltype(cuda::std::logb((__nv_bfloat16) 0)), __nv_bfloat16>::value), "");
+#endif // _LIBCUDACXX_HAS_NVBF16
+  assert(cuda::std::logb(1) == 0);
+}
+
+__host__ __device__ void test()
+{
+  test_log();
+  test_log10();
+  test_ilogb();
+  test_log1p();
+  test_log2();
+  test_logb();
+}
+
+__global__ void test_global_kernel()
+{
+  test();
+}
```

Comment:
Important: This is not called from anywhere, so the runtime `assert` is not checked.

Reply:
True, this is only to enforce instantiation in a __device__ context. We had some issues that nvcc sometimes does not surface usage of host only functions if they are not explicitly used within a __global__